Template/Trigger/ElementVisibilityTrigger.web.js

```diff
@@ -310,10 +294,86 @@
+                            if (node.contains(element)) {
+                                if (blockTrigger || (onlyOncePerElement && isNodeEventTriggered(element))) {
+                                    return;
+                                }
+
+                                if (!isNodeInViewport(element) && observerIntersection) {
+                                    observerIntersection.observe(element);
+
+                                    return;
+                                }
+
+                                var percentVisible = Math.max(getPercentVisible(element), minPercentVisible);
+                                commonTrigger(triggerEvent, percentVisible, element);
+                                if (fireTriggerWhen === 'oncePage') {
+                                    blockTrigger = true;
+                                    triggeredNodes.push(element);
+                                    if (self.scrollIndex) {
+                                        TagManager.window.offScroll(self.scrollIndex);
+                                    }
+                                } else if (onlyOncePerElement) {
+                                    triggeredNodes.push(element); // to avoid possible memory leaks as much as possible we add onceElement only when needed
+                                }
+                            }
+                        });
+                    });
+                }
+            }
+        }
+
+        function isNodeInViewport(node) {
+            var rect = node.getBoundingClientRect();
+
+            return (
+                rect.top >= 0 &&
+                rect.left >= 0 &&
+                rect.bottom <= (windowAlias.innerHeight || documentAlias.documentElement.clientHeight) && /* or $(window).height() */
+                rect.right <= (windowAlias.innerWidth || documentAlias.documentElement.clientWidth) /* or $(window).width() */
+            );
+        }
+
+        function commonTrigger(triggerEvent, percentVisible, node) {
+            var dom = TagManager.dom;
+            triggerEvent({
+                event: 'mtm.ElementVisibility',
+                'mtm.elementVisibilityPercentage': Math.round(percentVisible * 100) / 100,
+                'mtm.elementVisibilityId': dom.getElementAttribute(node, 'id'),
+                'mtm.elementVisibilityClasses': dom.getElementClassNames(node),
+                'mtm.elementVisibilityText': TagManager.utils.trim(node.innerText),
+                'mtm.elementVisibilityNodeName': node.nodeName,
+                'mtm.elementVisibilityUrl': node.href || dom.getElementAttribute(node, 'href')
+            });
+        }
+
         this.setUp = function (triggerEvent) {
-            this.scrollIndex = TagManager.window.onScroll(checkVisiblity(triggerEvent));
-            TagManager.dom.onLoad(checkVisiblity(triggerEvent));
+            if (!(isMutationObserverSupported && observeDomChanges && isIntersectionObserverSupported)) {
```

Comment:
@AltamashShaikh I think the code would be a more readable when writing eg 

```js
var useMutationObserver = isMutationObserverSupported && observeDomChanges && isIntersectionObserverSupported;
if (useMutationObserver) {
    TagManager.dom.onLoad(setMutationObserver(triggerEvent));
} else {
  this.scrollIndex = TagManager.window.onScroll(checkVisiblity(triggerEvent));
                TagManager.dom.onLoad(checkVisiblity(triggerEvent));
}
TagManager.dom.onLoad(setIntersectionObserver(triggerEvent));
```

When reviewing and looking at code it looked initially like mutation observer is always added etc.

Reply:
Yes makes sense, updated as suggested